yarn-project/sequencer-client/src/simulator/public_executor.ts

```diff
@@ -20,28 +20,37 @@ export function getPublicExecutor(
   contractDataSource: ContractDataSource,
   l1toL2MessageSource: L1ToL2MessageSource,
   blockData: HistoricBlockData,
+  newContracts: ExtendedContractData[] = [],
 ) {
   return new PublicExecutor(
     new WorldStatePublicDB(merkleTree),
-    new ContractsDataSourcePublicDB(contractDataSource),
+    new ContractsDataSourcePublicDB(contractDataSource, newContracts),
     new WorldStateDB(merkleTree, l1toL2MessageSource),
     blockData,
   );
 }
 
 /**
- * Implements the PublicContractsDB using a ContractDataSource.
+ * Implements the PublicContractsDB using a ContractDataSource and a set of new contracts.
  */
 class ContractsDataSourcePublicDB implements PublicContractsDB {
-  constructor(private db: ContractDataSource) {}
```

Comment:
If the public function call fails, we should remove the contract data for that transaction. Will need to modify the cache structure a bit for that I think.